File: python/wolvercote/serialise.py
"""
Serialise a CellSet back to a Wolvercote format string.
"""

from __future__ import annotations
from .types import Cell, CellSet, ChromosomeNode, MGENode, Replicon


def to_wolvercote(cell_set: CellSet) -> str:
    """Convert a CellSet back to a Wolvercote format string."""
    return " ; ".join(_cell_str(c) for c in cell_set.cells)


def _cell_str(cell: Cell) -> str:
    return ", ".join(_replicon_str(r) for r in cell.replicons)


def _replicon_str(r: Replicon) -> str:
    if isinstance(r, ChromosomeNode):
        inner = ", ".join(_mge_str(m) for m in r.children)
        attrs = _attrs_str(r.attributes)
        return f"({inner}){r.label}{attrs}"
    else:
        return _mge_str(r)


def _mge_str(m: MGENode) -> str:
    inner = ", ".join(_mge_str(c) for c in m.children)
    attrs = _attrs_str(m.attributes)
    return f"{{{inner}}}{m.label}{attrs}"


def _attrs_str(attrs: dict[str, str]) -> str:
    if not attrs:
        return ""
    pairs = ", ".join(f'{k}="{v}"' for k, v in attrs.items())
    return f"[{pairs}]"

File: webapp/src/wolvercote/serialise.ts
/**
 * Serialise a CellSet back to a Wolvercote format string.
 */

import type { Cell, CellSet, MGENode, Replicon } from './types'

export function to_wolvercote(cellSet: CellSet): string {
  return cellSet.cells.map(cellStr).join(' ; ')
}

function cellStr(cell: Cell): string {
  return cell.replicons.map(repliconStr).join(', ')
}

function repliconStr(r: Replicon): string {
  if (r.kind === 'chromosome') {
    const inner = r.children.map(mgeStr).join(', ')
    const attrs = attrsStr(r.attributes)
    return `(${inner})${r.label}${attrs}`
  }
  return mgeStr(r)
}

function mgeStr(m: MGENode): string {
  const inner = m.children.map(mgeStr).join(', ')
  const attrs = attrsStr(m.attributes)
  return `{${inner}}${m.label}${attrs}`
}

function attrsStr(attrs: Record<string, string>): string {
  const pairs = Object.entries(attrs)
  if (!pairs.length) return ''
  return '[' + pairs.map(([k, v]) => `${k}="${v}"`).join(', ') + ']'
}
